### ISSUE TITLE
approval-voting: coalesce multiple approvals in a single signed message

### ISSUE THREAD

**Issue by sandreim** [paritytech/polkadot#6831]
This is aimed at reducing the amount of messages and work done per relay chain block. Currently we send one signed message for each candidate approved under a block as soon as candidate validation completes the check. 

We should look at coalescing multiple approval votes for different candidates in a single message, instead of immediately sending individual messages out. This change would introduce some latency and we should calibrate the time we wait for more candidates to be approved to avoid no-shows. I expect this to work really well for tranche 0 approvals with https://github.com/paritytech/polkadot/pull/6782. In the happy case, each validator would send out exactly 2 messages per relay chain block for tranche0 checkers.

We would need to define some V3 primitives for [ApprovalVote](https://github.com/paritytech/polkadot/blob/master/primitives/src/v2/mod.rs#L1015) such that it wraps a bitfield in which each bit is represents the candidate by index in the inclusion relay chain block.

Also, [IndirectSignedApprovalVote](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/primitives/src/approval.rs#L106) wire message needs be updated to provide the signing context for verification.

**Comment by burdges** [comment]
We've discussed lengthening tranche time with a similar lengthening of the no-show time, which likely makes the timing here less tight.

Also, if we do not do that then we could've some extra information in the assignment VRFs which says when we expect to start work, and acts like delaying receiving some of the assignments within the VRF, so giving them a later no-show time.

**Comment by eskimor** [comment]
That one is going to be tricky with disputes, where we assume that there is one signed message per candidate/dispute. Also in order to verify the signature, we need to provide the [candidate hash](https://github.com/paritytech/polkadot/blob/126ca5819c2b7580c8cf601ccd4815ff20588baf/primitives/src/v4/mod.rs#L1251) for the signing payload. Currently we just pass the hash of the candidate in question for checking the signature. If we combine multiple, then suddenly we would need to know all the other candidate hashes as well.

I think this is solvable by reconstructing the signing payload (all signed candidate hashes) at the receiving side of the approval distribution and store it in the local representation of the vote. Some more compact representation might be possible as well. All we need is a proof that a given `CandidateHash` was in the approved set: But it should be self contained: Having to look up some relay chain block to gather indices in order to check the signature does not sound ideal.

Whether or not a bitfield is ideal is also a different question. If we assume only a couple of candidates to be checked by each validators (which should be the case), having a list of u16 indices might be more efficient and also more future proof:

1. The list can only grow as large as the maximum number of checks a single validator should do. This is a number that should stay more or less constant with regards to the number of parachains.
2. The list can grow freely, it does not matter whether it has 1 or 8 entries - message format does not change.
3. With a u16 we can have up to 65536 cores.

Size: 4xu16 = 8byte
"Future proof bitfield" >16 byte.

Depending on what we do with the above `CandidateHash` issue, we might end up not doing either, but instead just encoding the hashes themselves.

**Comment by sandreim** [comment]
Excellent point @eskimor . I actually did not consider disputes at the time of writing this :)

**Comment by burdges** [comment]
I think @alexggh results in https://github.com/paritytech/polkadot/issues/6608#issuecomment-1587081747 suggest the `Vec<CandidateHash>` optimization should be significant.

It's also a fine optimization when adversaries give us spam across multiple cores.

**Comment by sandreim** [comment]
Yes, this I also believe that, but it also depends on the other overheads aside from the sig checks like internal book keeping. Doing this would make processing approvals in parallel less efficient assuming that a work is split by candidate hash - workers would have to check signatures twice, but the other overhead should not be duplicated.

**Comment by burdges** [comment]
> Coalescing would make processing approvals in parallel less efficient assuming that a work is split by candidate hash - workers would have to check signatures twice, but the other overhead should not be duplicated.

Twice?  You're talking about a worker thread/future which processes approval (and assignment) gossip messages, and pushes them onto the tallying data structure, yes?  We'd need some extra `Mutex` which then creates blocking risks?  There exist data structures which minimize the blocking here, but yes they add complexity.

**Comment by burdges** [comment]
An easier fix, you cap the Vec lengths.

Alternatively, you could cap the approval votes messages from each sender within a time window, not sure if this demands uncapped Vecs or the Merkle thing.  This messes up our gossip assumptions, which sucks, but maybe if we used the sender's time stamp here then either their messages become somehow irrelevant, or else we merge these into a slashable equivocation in their approval votes.

Also, we're not talking about backers votes here.  I guess those contain more information.

**Comment by sandreim** [comment]
> > Coalescing would make processing approvals in parallel less efficient assuming that a work is split by candidate hash - workers would have to check signatures twice, but the other overhead should not be duplicated.
> 
> Why twice? You're talking about a worker thread/future which processes approval (and assignment) gossip messages, and pushes them onto the tallying data structure, yes? We'd need some extra `Mutex` which then creates blocking risks? There exist data structures which minimize the blocking here, but yes they add complexity.

Yeah I talk about having separate threads/workers that get assigned mutually exclusive per candidate work. A mutex could work here, to avoid the work if another thread has checked the signature before, but will defeat the purpose of parallelization of work, since we could block up to the time it takes to check a signature on a different worker. If we do not use the mutex, we need to check the signature twice if 2 workers need to look at the messages because they both have candidates to work on in the same message.

@eskimor you make an excellent point about duplicating storage, while yes, as @burdges suggests, capping how many approvals can we coalesce at a time would mitigate some of that. 

We could also add some additional constraints to how the coalescing works, such that the message itself is for candidates that will be checked by a single worker - assuming we define a function `f(candidate_hash) -> worker_id`. Of course we would need to check that the approval messages obey these new rules, thus more complexity.

**Comment by burdges** [comment]
As for data structures, if you're worried about stalling on mutexes, then at the extreme each candidate record could contain:
```
    incoming_approvals: [Mutex<LinkedList<(ValidatorId,CheckerStatus)>>; NUM_APPROVALS_THREADS],
```
As a signature checker thread, you append only to your own linked list, so they never compete for the `Mutex`.  About ever half second the tallier does 
```
let mut incoming_approvals = [LinkedList::new(), NUM_APPROVALS_THREADS];
for ia in incoming_approvals.iter_mut() {
    mem::swap( candidate.incoming_approvals[i].lock().deref_mut(), ia );
}
```
so they only hold each mutex for one swap instruction every half second or so.  It's now the talliers job to integrate each linked list into it's data structure before doing its tally.

These linked lists are not mmap friendly of course, even if the talliers data structures are made mmap friendly, so this limits syncing to disk, but that's not too bad.

---

We definitely have some complexity costs here so maybe we should double check how much this'll really save.

**Comment by sandreim** [comment]
I don't think the solution above can be applied to our already complicated codebase. 

The complexity goes really deep as we process approvals/assignments in 2 subsystems. In approval distribution we do some minimal processing to ensure deduplication of messages, avoiding working on finalized blocks as well as book keeping of individual messages via finger printing. All of this requires us to keep some state that we inspect in order to decide how we process an individual approval - for example, if we haven't seen an assignment before the approval we discard the approval (without importing it) and decrease the reputation of the peer that sent it.

However, the heavy lifting is done by approval voting when approval distribution requests it to import approvals and assignments. It's important to note that approval voting reads/writes to the DB, so having multiple workers requires that they work on differenet data sets. This makes sense wrt to the in memory state that we keep about the assignments and votes - 
own assignments, and no-shows, approvals seen already, tranches - such that the subsystem can know when to trigger own assignment and approval work. For all this to work nicely in parallel, candidates need to be processed exclusively by a single worker - all assignments and approval for that candidate have to be processed by same thread, written in the same DB without needing synchornization with other workers that would block and make things run in a serial fashion again.

That being said, the simplest way I see forward for implementing this parallel processing is to shard the state at the DB level by candidates, having the workers operate on different candidates in approval voting and enabling approval distribution to do the processing without waiting for individual approvals/assignments to be imported by approval voting workers.

I totally agree this complicates things some more, but it will be mostly plumbing work to avoid the serial processing of requests that we do now.

**Comment by eskimor** [comment]
Also worth reconsidering the lazy checking idea again, should be even faster and probably ends up being less complicated/invasive. Major blocker: Gossip concerns.

**Comment by burdges** [comment]
We'd need to check before sending the message back out again.  Abstractly I'd expect this adds complexity for little gain in the worst case (and the worst case matters)

> If we do not use the mutex, we need to check the signature twice if 2 workers need to look at the messages because they both have candidates to work on in the same message.

I suspect this breaks batch verification too, indeed it's likely harder.

Ignoring your specific subsystem divisions, we abstractly have roughly 4 phases in the pipeline:
1. Receive gossip messages over the wire and separate them from all the other message types
2. Check message validity
3. Do approval tallies 
4. Forward gossip messages as appropriate 

We've diverse concerns at boundaries between these phases, like 1-2 risks spam if messages accumulate unchecked, and 2 must know the forwarder to adjust reputation.  Yet, we cannot do any fancy signature optimizations if 2 does not happen independently first before 3 and 4. 

In particular, we cannot do batch verification for messages forwarded by multiple peers:  If Alice and Eve both send us messages, then we can batch verify the messages from each on separately.  If otoh we've some abstract verifier that batch verifies them both together then Eve can poison Alice's batch, and we must then reject everything before we touch reputation, so we've made our DoS problem worse, not better.  Also, this abstract batch verifier has its own internal mutexes, etc.

**Comment by burdges** [comment]
Yes, that's true, but..

We're seemingly not verifying them there, only later in the pipeline, where likely their separated by core.  We'd gain less by then because mostly they'd all land on different cores. 

I suppose our information dependency here is the candidate hash itself, so maybe the current system first checking the candidate hashes exist and splits them by candidate, yes?  I'm essentially saying this candidate hash check should only "attach" where the message goes, but not do the split until after the signature check happens.

**Comment by burdges** [comment]
If we do this then we should consider what happens when processing a vote in which some candidate hashes make sense, but not others.

We presumably now run politeness at the level of relay chain blocks now, no?  Assuming yes then we'd require approval votes always derive from only one specific relay chain block, maybe even include its block hash.  I think then little changes during the regular path, but..
- Are there new places where we should enforce this restriction?  I suspect no but not sure.
- Does anything change when two relay chain blocks contain the same candidate?

**Comment by burdges** [comment]
As both this and #7441 look complex, should we experiment with a faster signature anyways?

**Comment by alexggh** [comment]
> As both this and https://github.com/paritytech/polkadot/issues/7441 look complex, should we experiment with a faster signature anyways?

What do you mean by faster signatures, do you mean pulling the improvements mentioned here: https://github.com/paritytech/polkadot/issues/6608#issuecomment-1587339810  or using a different algorithm ? 

Another path, that I plan to experiment with is the lazy checking(suggested by @eskimor)  where we do not check the approvals at all the moment we received them and check them in the moment we have to import them on chain, but that has some other implication with the gossip algorithm, which I'm currently trying to fully understand. On the best case, we would have to not gossip at all and ask each validator to send its approval to all others.

**Comment by alexggh** [comment]
## Summary
Alright, based on the feedback and data I've got in https://github.com/paritytech/polkadot/pull/7482, https://github.com/paritytech/polkadot/pull/7393 and other discussions we had inside the team, I think this optimisation is what should implement next and this together with https://github.com/paritytech/polkadot/pull/6782 should gives us enough gains, so that the approvals-voting & approvals-distribution, not be the bottleneck when increasing the numbers of validators and parachains.

## Addressing concerns

### Why not `core groups` https://github.com/paritytech/polkadot/issues/7441 ?
The idea should give us gains across the stack, however that has a longer term horizon, since it would require changes everywhere. However, we plan to implement the coalescing of approvals in a generic way, so it can be used for core-groups as well, so this work can be seen as sub-task for implementing core-groups.

### Why give up on no-gossip of approvals + lazy signature checks https://github.com/paritytech/polkadot/pull/7482 ?
The idea has some merits and could theoretically give us better gains, however it a few major drawabacks:
- Overall it made things more fragile(finality could stall accidentally) and could compromise the security of the network, because of the added complexity, see https://github.com/paritytech/polkadot/pull/7482#issuecomment-1642037387
- It locked us on path where: https://github.com/paritytech/polkadot/issues/7229 would be possible, without some other complex solutions.

### Other ideas considered
#### Perform just parallelization: https://github.com/paritytech/polkadot/pull/7393
This will gives us some gains, but they are definitely lower than reducing the amount of work with coalescing, however if need be, we can implement some sort of parallelization of the crypto work in the unlikely case where we would still need gains after the coalescing approvals work.

####  Approval votes by hash inversion vs faster assignment VRF https://github.com/paritytech/polkadot/issues/7543
@burdges doesn't endorse it either, because of the added complexity and its downsides, so I don't think we should go this route either.

### Proposed approach.
1. Vote for multiple candidate with a single signatures.
2. The number of candidates we can included will be bounded. The obvious bound could be `relayVrfModuloSamples`, but we can define a separate configuration parameter if we want to decouple it.
3. Change [IndirectSignedApprovalVote](https://github.com/paritytech/polkadot/blob/master/node/primitives/src/approval.rs#L105C7-L105C7) to reflect the fact that we signed more than one candidate. To be consistent with https://github.com/paritytech/polkadot/pull/6782, it will be a bitfield.
4.  For disputes when we have to store the votes on chain we will have to modify [DisputeStatementSet here](https://github.com/paritytech/polkadot/blob/master/primitives/src/v5/mod.rs#L1356), to include a `Vec<CandidateHash>`. This has the downside it that would increase the necessary storage requirements by 192 bytes `relayVrfModuloSamples * size_of(CandidateHash)`. Given disputes are not the happy path and should be rare, I think that increase shouldn't be deal breaker.


### Other considerations
 #### Time disputes implementation https://github.com/paritytech/polkadot/issues/6510
The timing information for all candidates included in the signature would have to be imported on-chain, but assuming it is just an `u32` and the fact the number of candidates we can sign is small, the necessary storage shouldn't be a problem.

@burdges @sandreim @eskimor @ordian let me know what you think about this, meanwhile I'll go ahead and implement a PoC on top of https://github.com/paritytech/polkadot/pull/6782

**Comment by sandreim** [comment]
> ## Addressing concerns
> ### Why not `core groups` #7441 ?
> The idea should give us gains across the stack, however that has a longer term horizon, since it would require changes everywhere. However, we plan to implement the coalescing of approvals in a generic way, so it can be used for core-groups as well, so this work can be seen as sub-task for implementing core-groups. 
> 

Also I think these could stack. Once core groups implemented, assignments/approvals will refer to a core group index. A bitfield representing the signed core groups or a vec of core group indices (assuming u8). It will however require to get know all candidate hashes when checking the signature on node side and runtime.

> ### Proposed approach.
> 1. Vote for multiple candidate with a single signatures.
> 2. The number of candidates we can included will be bounded. The obvious bound could be `relayVrfModuloSamples`, but we can define a separate configuration parameter if we want to decouple it.

We also need to implement some constraints here like maximum wait time until we sign approval and gossip such that it doesn't interfere with finality. I'd have two different configuration options: `maxApprovalCoalesceCount` which refers to candidate count, `maxApprovalCoalesceWait` in relay relay chain blocks which should be less than no show slots.

I
FWIW, this sounds like a good plan.

**Comment by alexggh** [comment]
Added an initial draft of the proposal here: https://github.com/paritytech/polkadot/pull/7554, feedback welcomed.
